Simplify static forms (#732)
ghc-proposals/ghc-proposals: proposals/0000-simplify-static.rst

Simplifying static forms
==============

.. author:: Simon Peyton Jones
.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
.. ticket-url:: Leave blank. This will eventually be filled with the
                ticket URL which will track the progress of the
                implementation of the feature.
.. implemented:: Leave blank. This will be filled in with the first GHC version which
                 implements the described feature.
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/732>`_.
.. sectnum::
.. contents::

This proposal simplifies the typing rules for ``static`` forms. The current design

> is too elaborate and the implementation never worke.d

Reviewer: ```suggestion
is too elaborate and the implementation never worked.
```

> is too elaborate and the implementation never worke.d

Reviewer: ```suggestion
is too elaborate and the implementation never worked.
```

Motivation
----------

The ``static`` form in GHC is `described in GHC's user manual <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/static_pointers.html>`_. It allows you to define a "static pointer", of type ``StaticPtr a``.

In the form ``(static e)`` the intuitive idea is that ``e`` is "pure code", with no free variables;
that is, all ``e``'s free variables are bound at top level.  For example::

  y :: String
  y = "hello" ++ " Fred"

  f x = ....(static (reverse y))...

But GHC also allows this::

  f x = let { y = "hello" ++ " Fred" } in static (reverse y)

> fHere `y` is bound nested-ly, but is morally top-level because it in turn has no

Reviewer: ```suggestion
Here `y` is bound nested-ly, but is morally top-level because it in turn has no
```

> fHere `y` is bound nested-ly, but is morally top-level because it in turn has no

Reviewer: ```suggestion
Here `y` is bound nested-ly, but is morally top-level because it in turn has no
```

the ``static`` is not just code: it mentions a dynamically bound (albeit invisible) parameter to ``f``.

Solution: wrap the constraints in an implication with ``SkolInfo`` of ``StaticFormSkol``; and
in the constraint solver zap all Givens when walking inside such an implication. That was
done in::

  commit 39d4a24beaa7874a69ffdc1528ca160818829169
  Author: Simon Peyton Jones <simon.peytonjones@gmail.com>
  Date:   Tue Sep 30 23:11:19 2025 +0100

    Build implication for constraints from (static e)

    This commit addresses #26466, by buiding an implication for the
    constraints arising from a (static e) form.  The implication has
    a special ic_info field of StaticFormSkol, which tells the constraint
    solver to use an empty set of Givens.

So that complication was unwelcome, but not too bad.

**Complication 2**.  What if we have::

  f x = let y = reverse "hello" in
        ...(static (y++y))...

The free variables of the static are just ``{y}``, and ``y`` is "morally-top-level": its right-hand side
has no free variables.
Sadly (as it turns out) GHC tries to accept this case.  When looking at the defn of y

> (with no static in sight yet) the typechecker marks it at a "static binding", meaning that

Reviewer: ```suggestion
(with no static in sight yet) the typechecker marks it as a "static binding", meaning that
```

Proposed Change Specification
-----------------------------

I propose the following change, which returns to the original spec of `static e`:

* In an expression ``static e``, the free term variables of ``e`` must all be bound at top level.

I propose no change to the existing (but un-stated) restrictions:

* In an expression ``static e`` all typing constraints arising from ``e`` must be soluble using global instance declarations only.  (This is Complication 1; it is an existing rule which is unaffected by this proposal.)
* In an expression ``static e``, where ``e :: ty``, the type ``ty`` must satisfy ``Typeable ty``

That is the complete specification.

I propose that we make this change **without** a deprecation cycle, for the following reasons:

* It is hard to fix currently-open bugs with the current specification, as described above.
  Retaining these bugs for another year while we deprecate ``static`` with free nested variables seems undesirable.

* I konw of no libraries that would be adversely affected.  That's not to say they don't exist, of course.

> * Even if a library is affected, the fix is very easy.

Reviewer 2: I disagree with this. I don't think this should set a precedence for skipping deprecation cycles.

1. It sets a bad precedent.
2. I don't see retaining bugs due to backwards compatibility necessarily a problem? If it works today (by accident), why break existing working code?
3. There is no reason to try to _fix_ the existing deprecated implementation. It's deprecated, users are encouraged to move to a new implementation?
4. A code-level fix is _never easy_. Of course it's easy at the site where it is, but downstream? Now you need a new library release for the affected library, and you need a potentially new release for any intermediate libraries, until your own indirectly affected code can compile with a newer compiler. That is a lot of churn. (Also the semantics of the original library might now change if you end up depending on an older minor version; and the fix will only be available in a newer major version with different behaviour.)

Proactively patching packages will not be always possible, there are maintainers that refuse to update their packages until a proper release is cut. Arguably there is a good reason for this, who knows if what ever was updated really ends up in a release, and if not one has put some half-baked release out.

I have no idea what the real world impact is. I am not sure that has any impact though, breakage is breakage; and I can't support abrupt breakage.

Author: Sadly, not easy.

The current situation is this
* Rather late in the optimisation chain, the "float-out" pass has an ad-hoc reocogiser that spots applications of `static`, and floats them to the top level.
* This is fragile because it relies on the application `(static e)` not geting separated or confused. It also relies on the free variables of `e` also getting floated.
* There is other ad-hoc stuff in the back end.

My patch radically simplifies this story -- that's the point.
* The desugarer, which converts from `HsExpr` to `CoreExpr`  immediately promotes `HsStatic e` to top level.
* All the ad-hoc junk later in the compiler disappears.

Sadly it is hard to use the new plan if there are nested let-bindings that are free in the `HsStatic e` form.  They would
need to be promoted too somehow.  But really only the ones that appear free in an inner `HsStatic`. Implementing that would be a pain, especially since it'd be discarded entirely
when the proposal is implemented.

If we want a deprecation cycle I think we have to do the following:
* Put [my patch](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/14994) on ice.
* Instead, keep the existing implementation, bugs and all, unchanged.  Mark the tickets as won't fix until the patch lands.
* Add a new patch that complains if a `static` has a free variable -- this is the deprecation warning.  Perhaps this could go into 9.14.
* Then (I guess in 9.16) implement the new behaviour.

Would a one-release deprecation cycle be acceptable @angerman? Now I think about it, if we got the deprecation patch into 9.14, we can implement the new behaviour in HEAD and it'll show up in 9.16.  So that's not so bad.

Reviewer 2: 9.14 -> 9.16 is certain better than nothing.

Is there a reason why we can't accept that `StaticPointers` as it is is simply broken, and will be deprecated outright as a language extension, and we'll have a new `StaticPointers2` language extension that has different behaviour, design, ...?

Author: > Is there a reason why we can't accept that StaticPointers as it is is simply broken, and will be deprecated outright as a language extension

Yes I think there is: doing so would be a breaking change.  Some users rely on `StaticPointers` and it seems unhelpful simply to remove them.

If you or any one else wants to make a GHC proposal about an alternative, the path to do so lies open.  But _this_ proposal is about a much narrower question: making a small but significant  simplication to an existing feature that, a change that will impact very few users, will materially simplify the implementation, and will fix reported bugs.

Reviewer 3: This sounds reasonable, but I feel slapping numbers on sets a bad precedent. Maybe we could name it `TopLevelStaticPointers` instead, which also hints at the restriction.

Reviewer 4: I don't see that a new extension would give much benefit, and it imposes significant costs, notably that existing users of `StaticPointers` whose code already satisfies the restriction would have to take action to explicitly ask for the non-buggy version.

If it is feasible to implement both the old and new behaviour side-by-side, as would be required for supporting both old and new extensions, then couldn't you just as well have a single `StaticPointers` extension that decides for each `static e` expression whether to use the old or the new codepath based on whether it is acceptable under the new rules (i.e. whether all its free variables are bound at the top-level)?

This leaves open the question of whether this is feasible at all. @simonpj argues in https://github.com/ghc-proposals/ghc-proposals/pull/732#discussion_r2590371853 that it is difficult to support both at once, although I still don't quite see why. Simon, you write:

> Sadly it is hard to use the new plan if there are nested let-bindings that are free in the `HsStatic e` form. They would need to be promoted too somehow.

To be clear, my suggestion is that at an early stage the typechecker (presumably) would test whether any such problematic free occurrences exist, and tag the `HsStatic` as subject to the old or new plan based on this. So code that works under the new plan gets the non-buggy treatment, and code that doesn't gets a warning but continues to use the old route. I see that this requires keeping around a bunch of old code in float-out that we'd like to get rid of eventually, but are there other complications I'm not seeing?

Reviewer 2: I see this as follows:
- Option 1: break immediately and abruptly. New compiler -> broken code.
- Option 2: Some form of immediate deprecation warning, and abrupt breakage in the next compiler -> short warning period; some package might not upgrade until a new compiler is officially released.
- Option 3: Acknowledging that the StaticPointer specification is ill specified, and the extension needs to be changed. New extension in parallel to the old extension, lots of flexibility around deprecation timeframes (both extensions can exist at the same time!).

Changing the behaviour of existing extensions with compilers means that the behaviour of extensions is a function of time. Thus examples, stack-overflow answers, documentation, books, ... written (e.g. basically anything frozen in time), will have to adapt.

To me the honest path forward is to accept that the extension as specified is bad; and needs to be overhauled, this should imo be a revision. What we are proposing to do is to do an in-place update with the associated deprecation complications.

There are several functions in this API which expect static pointers of dictionaries, say of the form::

  mkCommand :: Typeable a => StaticPtr (Dict (Binary a, Eq a)) -> ...

  -- NB: Dict is defined as follows:
  data Dict c where { Dict :: c => Dict c }

This means the code users write typically looks like::

  myPreBuildRules :: PreBuildComponentInputs -> RulesM ()
  myPreBuildRules = ... cmd1 ... cmd2 ...
    where
      cmd1 = mkCommand (static Dict) ...
      cmd2 = mkCommand (static Dict) ...

This code is fine: the body of these `static` calls is just ``Dict`` and has no free variables.
(As now, the rules require that the typing constraint arising from the use of ``Dict``
are top-level-soluble.)  In short, no back-compat issues arise here. (Thanks to @sheaf for summarising.)

Impact on Hackage
~~~~~~~~~~~~~~~~~~

I used `Hackage search <https://hackage-search.serokell.io/?q=StaticPointer>`_ to search for all
Hackage packages that mention "StaticPointer".  Therse were 51 such packages.  I looked at almost

> all of them.  I only one (namely ``imprint``) did I find a use of ``static`` which disobeyed

Reviewer: ```suggestion
all of them.  In only one (namely ``imprint``) did I find a use of ``static`` which disobeyed
```

  hindent-6.2.1
    Code layout

  hint-0.9.0.8
    Hints only

  hinter-0.1.0.0
    Hints only

  hlint-3.10
    Linting only

  homplexity-0.4.8.1
    Spurious

  hs-aws-lambda-0.2.0.0
    StaticPointers but no `static`!

  imprint-0.0.1.0
    One use, in a test only
      funImprint :: Imprint 'Z (Int -> String -> String)
      funImprint = I.static (static f)
        where
          f n str = str ++ show n

  jvm-0.6.0
    StaticPointers but no `static`!

  jvm-batching-0.2.0
  jvm-streaming-0.4.0
  matterhorn-90000.1.1

  ormolu-0.8.0.2
    Code formattting only

  porcupine-core-0.1.0.1
    StaticPointers but no `static`!

  rio-process-pool-1.0.1
  rio-0.1.24.0
    Spurious

  skylighting-0.14.7
  skylighting-core-0.14.7
    Spurious

  sparkle-0.7.4
    A few statics all fine

  static-closure-0.1.0.0
    Only one `static`, fine

  unliftio-messagebox-2.0.0
    Spurious

Alternatives
------------
The status quo is not really an alternative: we have bugs that I don't know how to fix.
I suppose that with yet more effort we could do something that never crashed the compiler;
but was merely hard to explain to the user.  But I think a more likely outcome is that

> we'll just have un-fixed bugs, and a hard-to-understand implementation, indefinitely.

Reviewer 5: I think a missing alternatives, which at least deserves an argument to be dismissed is: asking all the free variables to be static. Something like

```haskell
  f x = let y = static "hello" in   -- Not OK: y is not bound at top level
        static (derefStaticPtr y ++ derefStaticPtr y)
```
